core/src/main/java/org/apache/iceberg/avro/PruneColumns.java

```diff
@@ -144,16 +144,18 @@ public Schema array(Schema array, Schema element) {
       if (selectedIds.contains(keyId) || selectedIds.contains(valueId)) {
         return complexMapWithIds(array, keyId, valueId);
       } else if (element != null) {
-        Schema keyProjection = element.getField("key").schema();
+        Schema.Field keyProjectionField = element.getField("key");
```

Comment:
I wonder we should solve this in a different way. The issue is that when complex value types are involved for logical maps the key is not projected. Then why not we project the key field if there's a value field is projected.

Would it be better to change the `map` method in `GetProjectedIds` to always project the `keyField` if `valueResult` is not `null`? 

What do you guys think @rdblue, @lxynov ?

Reply:
Discussing with @rdsr offline, also posting here for @rdblue 's visibility

I tried to always project the `keyField` but encountered two issues:
 - `TestGenericReadProjection` in Parquet fails somehow. (I haven't spent time looking into how to resolve it)
 - We'll still need to change `PruneColumns` because its current logic is `// if either key or value is selected, the whole map must be projected`.

@rdsr @rdblue what are your thoughts? I think we can make it to always project `keyField` but it may need some efforts

Reply:
I understand now, specially bullet item 2 makes sense to me.